File: lib/services/nostr_keys.dart
import 'dart:typed_data';

import '../src/rust/api/nostr_keys.dart' as rust;
import 'settings.dart';

/// Generador/convertidor de claves Nostr — porte del `Keyl` de Gtool.
///
/// Todas las llamadas al codegen FRB son async.
class NostrKeys {
  /// Clave privada aleatoria (32 bytes).
  Future<Uint8List> generate() async =>
      Uint8List.fromList(await rust.nostrGenerateKey());

  /// Clave determinística desde semilla textual (SHA256 → 32 bytes).
  Future<Uint8List> fromSeed(String seed) async =>
      Uint8List.fromList(await rust.nostrSeedToKey(seed: seed));

  /// Bytes del secreto → nsec bech32 ('' si inválido).
  Future<String> toNsec(Uint8List secret) => rust.nostrToNsec(secret: secret);

  /// Bytes del secreto → npub bech32 ('' si inválido).
  Future<String> toNpub(Uint8List secret) => rust.nostrToNpub(secret: secret);

  /// npub (bech32 o hex) → hex.
  Future<String> hexNpub(String npub) => rust.nostrHexNpub(npub: npub);

  /// nsec (bech32 o hex) → hex.
  Future<String> hexNsec(String nsec) => rust.nostrHexNsec(nsec: nsec);

  /// Valida y normaliza a bech32. '' = inválido.
  Future<String> validateNpub(String input) =>
      rust.nostrValidateNpub(input: input);

  Future<String> validateNsec(String input) =>
      rust.nostrValidateNsec(input: input);

  /// Hex de la pública desde bytes del secreto.
  Future<String> pubkeyHexFromSecret(Uint8List secret) =>
      rust.nostrPubkeyHexFromSecret(secret: secret);

  // ---------------- Identidades guardadas (cifrado en config.pr) --------

  /// Genera identidad nueva con nombre; retorna {npub, nsec}.
  Future<Map<String, String>> createIdentity(String nombre, {String? seed}) async {
    final secret = seed == null ? await generate() : await fromSeed(seed);
    final npub = await toNpub(secret);
    final nsec = await toNsec(secret);
    await saveIdentity(nombre, secret);
    return {'npub': npub, 'nsec': nsec};
  }

  Future<void> saveIdentity(String nombre, Uint8List secret) async {
    final s = Settings.instance;
    s.nostrKeys.removeWhere((k) => k['nombre'] == nombre);
    s.nostrKeys.add({
      'nombre': nombre,
      'secretHex': _toHex(secret),
      'npub': await toNpub(secret),
      'createdAt': DateTime.now().millisecondsSinceEpoch,
    });
    await s.save();
  }

  List<Map<String, dynamic>> listIdentities() => Settings.instance.nostrKeys;

  Uint8List? loadSecret(String nombre) {
    for (final k in listIdentities()) {
      if (k['nombre'] == nombre && '${k['secretHex']}'.isNotEmpty) {
        return _fromHex('${k['secretHex']}');
      }
    }
    return null;
  }

  Future<bool> deleteIdentity(String nombre) async {
    final s = Settings.instance;
    final before = s.nostrKeys.length;
    s.nostrKeys.removeWhere((k) => k['nombre'] == nombre);
    await s.save();
    return s.nostrKeys.length != before;
  }
}

String _toHex(Uint8List b) =>
    b.map((x) => x.toRadixString(16).padLeft(2, '0')).join();

Uint8List _fromHex(String hex) {
  final out = Uint8List(hex.length ~/ 2);
  for (var i = 0; i < out.length; i++) {
    out[i] = int.parse(hex.substring(i * 2, i * 2 + 2), radix: 16);
  }
  return out;
}

File: rust/src/api/nostr_keys.rs
/// Generador de claves Nostr — portado de Gtool `keytool.rs` sin Godot.
use nostr::prelude::*;
use std::str::FromStr;

/// Clave privada aleatoria (32 bytes).
#[flutter_rust_bridge::frb]
pub fn nostr_generate_key() -> Vec<u8> {
    let keys = Keys::generate();
    keys.secret_key().secret_bytes().to_vec()
}

/// Clave derivada de una semilla textual (SHA256 → 32 bytes).
#[flutter_rust_bridge::frb]
pub fn nostr_seed_to_key(seed: String) -> Vec<u8> {
    use sha2::{Digest, Sha256};
    let mut hasher = Sha256::new();
    hasher.update(seed.as_bytes());
    hasher.finalize().to_vec()
}

/// Bytes de clave privada → nsec (bech32). Vacío si son inválidos.
#[flutter_rust_bridge::frb]
pub fn nostr_to_nsec(secret: Vec<u8>) -> String {
    let Some(sk) = secret_from_bytes(&secret) else { return String::new() };
    let keys = Keys::new(sk);
    match keys.secret_key().to_bech32() {
        Ok(nsec) => nsec,
        Err(_) => String::new(),
    }
}

/// Bytes de clave privada → npub (bech32). Vacío si son inválidos.
#[flutter_rust_bridge::frb]
pub fn nostr_to_npub(secret: Vec<u8>) -> String {
    let Some(sk) = secret_from_bytes(&secret) else { return String::new() };
    let keys = Keys::new(sk);
    match keys.public_key().to_bech32() {
        Ok(npub) => npub,
        Err(_) => String::new(),
    }
}

/// npub (bech32 o hex) → hex. Vacío si es inválido.
#[flutter_rust_bridge::frb]
pub fn nostr_hex_npub(npub: String) -> String {
    match PublicKey::from_bech32(&npub) {
        Ok(pk) => pk.to_string(),
        Err(_) => String::new(),
    }
}

/// nsec (bech32 o hex) → hex. Vacío si es inválido.
#[flutter_rust_bridge::frb]
pub fn nostr_hex_nsec(nsec: String) -> String {
    match SecretKey::from_bech32(&nsec) {
        Ok(sk) => sk.to_secret_hex(),
        Err(_) => String::new(),
    }
}

/// Valida y normaliza un npub (acepta bech32 o hex, devuelve bech32).
#[flutter_rust_bridge::frb]
pub fn nostr_validate_npub(input: String) -> String {
    if let Ok(pk) = PublicKey::from_bech32(&input) {
        return pk.to_bech32().unwrap_or_default();
    }
    if let Ok(pk) = PublicKey::from_str(&input) {
        return pk.to_bech32().unwrap_or_default();
    }
    String::new()
}

/// Valida y normaliza un nsec (acepta bech32 o hex, devuelve bech32).
#[flutter_rust_bridge::frb]
pub fn nostr_validate_nsec(input: String) -> String {
    if let Ok(sk) = SecretKey::from_bech32(&input) {
        return sk.to_bech32().unwrap_or_default();
    }
    if let Ok(sk) = SecretKey::from_str(&input) {
        return sk.to_bech32().unwrap_or_default();
    }
    String::new()
}

/// Hex de la clave pública a partir de bytes del secreto.
#[flutter_rust_bridge::frb]
pub fn nostr_pubkey_hex_from_secret(secret: Vec<u8>) -> String {
    let Some(sk) = secret_from_bytes(&secret) else { return String::new() };
    Keys::new(sk).public_key().to_string()
}

fn secret_from_bytes(secret: &[u8]) -> Option<SecretKey> {
    let bytes: [u8; 32] = secret.try_into().ok()?;
    SecretKey::from_slice(&bytes).ok()
}
